[team-09][Lollo, ZG] 데이터 받아 테이블 뷰 표시하기

Settpark opened the pull request:
### 작업 내용

1. MVVM의 구조로 설계하도록 노력했습니다.
2. Network의 내부 함수를 한가지 기능만 할 수 있게 독립적으로 나누려고 노력했습니다.
3. MVVM의 구조에서 계층 간의 의존 관계를 끊으려고 하였습니다.
4. Mock API를 통하여 메인화면의 일부 데이터를 받아와서 테스트 하였습니다.

### 학습 키워드
1. Network + Combine
2. MVVM
3. Clean Architecture

### 고민 & 질문 사항
1. UseCase를 다루는 형태를 고민했습니다. UseCase를 알고 있어야 하는 대상의 구분, UseCase의 역할 등을 고민하였고
현재 UseCase가 잘 쓰이고 있는 지에 대한 고민이 가장 컸습니다. -> 완전히 해결하지 못한 상태로 남아있습니다.
2. 현재 ViewModel에 sideDishesCollection의 구체 타입을 알고 있습니다. 구조상 뷰모델에서 도메인의 entities를 알고 있는 것에 대한 고민이 있는 데, 이 것을 단지 프로토콜로 전환한다고 해서 의존성이 끊어지는 지에 대한 의문이 남아 현재 진행하지 않고 구체 타입을 알고 있습니다. 프로토콜로 개선하면 이 문제를 해결할 수 있을까요??
3. Network 모델 설계 시 재사용성을 고민하였습니다. Network 객체의 url을 외부에서 지정해줄 수 있도록 구현하고, Request 생성 객체 메소드에서 Enum 타입의 RequestType과 EndPoint를 넘겨줄 수 있도록 했습니다. EndPoint의 경우 hash 값 등 반찬 cell 선택에 따른 값 설정이 필요할 수 있어서 추후 API 형식에 따른 타입 변경이 필요해보입니다.
4. Repository의 역할을 구분하는 기준이 명확하지 않아 현재의 프로젝트에서 repository의 영역을 구현하지 못했습니다. 혹시 현재 프로젝트에서 어떠한 방식으로 repository를 구현해야 하는지, 방향성에 대해서 질문 드리고 싶습니다.

Commit:
Initial commit

Settpark committed:
style: 메인 화면 UI 구성 (#8)

TableView와 기본 뼈대가 되는 Cell을 Storyboard에 추가

Settpark committed:
feat: tableView의 customcell class 생성 (#8)

label 업데이트 함수 작성. 후에 수정 필요

Settpark committed:
feat: mainTableView DataSource 지정 (#8)

테이블 뷰와 DataSource 연결

Settpark committed:
feat: mainTableView의 Delegate 지정 (#8)

테이블 뷰와 delegate 연결 커스텀 헤더 지정

eeeesong committed:
Merge pull request #13 from ghojeong/feature/mainviewUI

Feature/메인 화면 UI 구성 (#8)

Settpark committed:
style: subdetailview 구성 (#9)

버튼 및 취소선 추가구성 필요

eeeesong committed:
feat: SideDish Entity 생성

- SideDish 및 Detail 구현
- 구조에 대한 논의 필요

eeeesong committed:
feat: SideDish Entity 생성

- SideDish 및 Detail 구현
- 구조에 대한 논의 필요

eeeesong committed:
feat: SideDishes Entity 생성

- SideDish 단일 리스트를 관리하는 모델 생성
- 추후 의존성 약화 위해 프로토콜로 인터페이스 분리 필요

eeeesong committed:
feat: SideDishes Entity 생성

- SideDish 단일 리스트를 관리하는 모델 생성
- 추후 의존성 약화 위해 프로토콜로 인터페이스 분리 필요

Settpark committed:
fix: 피드백 사항 개선 (#8)

headerView의 백그라운드색 지정, init함수 수정

eeeesong committed:
Merge pull request #19 from ghojeong/feature/subviewUI

Feature/subview UI

Settpark committed:
Merge branch 'feature/sidedishEntity' of https://github.com/ghojeong/sidedish into feature/sidedishEntity

eeeesong committed:
feat: SideDishesCollection Entity 생성

- SideDishes들의 리스트를 관리하는 모델 생성
- recentUpdate를 통해 데이터 버전이 관리하도록 구성

eeeesong committed:
feat: SideDishesCollection Entity 생성

- SideDishes들의 리스트를 관리하는 모델 생성
- recentUpdate를 통해 데이터 버전이 관리하도록 구성

eeeesong committed:
feat: SideDishes Entity 사용성 개선

- Category enum 생성하여 카테고리 설정 안정화
- SideDishFindable 인터페이스 설정하여 의존성 개선

eeeesong committed:
feat: SideDishes Entity 사용성 개선

- Category enum 생성하여 카테고리 설정 안정화
- SideDishFindable 인터페이스 설정하여 의존성 개선

Settpark committed:
chore: MVVM에 기초하여 폴더 구성 수정함

Settpark committed:
chore: MVVM에 기초하여 폴더 구성 수정함

Settpark committed:
Merge branch 'feature/sidedishEntity' of https://github.com/ghojeong/sidedish into feature/sidedishEntity

Settpark committed:
chore: 데이터 형식 확정

Settpark committed:
chore: 데이터 형식 확정

Settpark committed:
Merge pull request #21 from ghojeong/feature/sidedishEntity

너무 고생했습니다 :) (#11)

eeeesong committed:
feat: Network Model 생성 (#23)

- 작업 단위 별로 객체를 분리하여 구현

Settpark committed:
Merge branch 'feature/networkModel' of https://github.com/ghojeong/sidedish into feature/ViewModelBinding

Settpark committed:
feat: ViewModel과 ViewController의 binding

Settpark committed:
chore: 함수명 collection, dishes의 접근 수준 변경

eeeesong committed:
chore: Network 구조 개선

Settpark committed:
Merge pull request #28 from ghojeong/feature/networkModel

너무 고생했어요 :)

napster-x-90 commented on ios/sidedish/sidedish/Domain/Entity/SideDish.swift:
```diff
@@ -0,0 +1,28 @@
+
+import Foundation
+
+class SideDish: Codable {
```
Codable을 사용한 이유가 무엇인가요?
단순하게 서버의 JSON을 Decoding  하는 거라면, Decodable만 채택해서 사용하는게 좋습니다.
Codable을 쓴다면 해당 객체를 가지고 Encoding작업도 하는건가? 라는 생각을 할 수 있습니다.
명확하게 필요한 것만 채택해주시면 좀더 깔끔한 코드가 될꺼에요. 😄

해당 클래스에 전체적인 리뷰를 드리자면,
1. Swift Style Guide에 맞게 변수명을 지어주세요. (https://swift.org/documentation/api-design-guidelines/)
보통 `var detailHash` 이렇게 지어주세요. _ 로 구분을 하는걸 선호하지는 않습니다. 

2. n_price, s_price 등등 좀더 명확한 이름이 필요합니다! 주석이 없다면 어떤 역활을 하는 Property인지 전혀 알 수가 없습니다. 물론 여기선 서버에서 내려주는 거니 이렇게 하신거겠죠? JSON을 디코딩하기위해서 카멜케이스를 이용할 수 있는 옵션이 있습니다. 한번 찾아보세요.

3. SideDish를 Class로 만드신 이유가 있을까요? Struct로 만들면 안되는 이유라도 있을까요?

4. 속성이 배열이라면 s 를 붙혀주셔야 합니다.

다른 클래스들도 한번 확인해보세요!

eeeesong replied:
그렇군요! 명확하게 표현할 수 있도록 Decodable로 수정하도록 하겠습니다.
이 클래스의 경우 Mock API에 제공된 JSON의 key 값을 그대로 따오다 보니 Swift 컨벤션에 맞지 않는 변수명을 사용하게 되었습니다.😅 (그렇더라도 CodingKeys를 사용할 수 있다는 걸 간과한 것 같네요...)
그리고 초기 설계에서는 내부 프로퍼티를 변경할 일이 있어서 Class로 만들었는데, 지금 형태에서는 Struct로 만들면 안 될 이유가 없어 보입니다. 부주의했던 부분인 것 같네요! 짚어주셔서 감사합니다.

eeeesong replied:
피드백에 따라 수정을 마쳤습니다.

- Swift Style Guide에 따라 바꾸고, CodingKey를 추가하였습니다.
- Struct로 개선하였습니다.
- 배열 속성엔 복수형 네이밍을 사용하였습니다.

Settpark committed:
feat: Toast 추가

napster-x-90 commented on ios/sidedish/sidedish/Domain/Entity/SideDish.swift:
```diff
@@ -0,0 +1,28 @@
+}
+
+class Detail: Codable {
+    var thumbImages: [String] //library/cache의 파일명들
```
여긴 잘지켜져 있네요 :)
다만 전체적으로 내부 속성이 var로 선언하면 변경의 위험이 있으니, let으로 선언해 주시기 바랍니다.

napster-x-90 commented on ios/sidedish/sidedish/Domain/Entity/SideDishes.swift:
```diff
@@ -0,0 +1,28 @@
+    
+    private var category_id : String
+    private(set) var name : String
+    private(set) var items : [SideDish]
```
이미 `SideDishFindable` 에서 `func sideDish(at index: Int) -> SideDish?` 로 items를 read할 수 있는데
private(set)이 꼭 필요할까요?? 만약 필요하다면 `func sideDish(at index: Int) -> SideDish?` 이 오퍼레이션의 존재가 무엇일까요?

----> 추가 리뷰
저기 멀리에서 items[section].count로 사용하고 있었군요!

napster-x-90 commented on ios/sidedish/sidedish/Domain/UseCase/TurnonAppUsecase.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+    
+    convenience init(){
+        let networkmanager = NetworkManager(baseAddress: "https://h3rb9c0ugl.execute-api.ap-northeast-2.amazonaws.com")
```
Mock 객체를 초기화할때도 baseAddress를 주입받는 식으로 바꿔보세요.
해당 클래스를 테스트하기위해 직접 생성하는건 다른 주소의 base를 넣고 싶을땐 할 수가 없게 됩니다.

```swift
init(networkmanger: NetworkProtocol = NetworkManager(baseAddress: "www.apple.com")) {
...
}
```
이런 방법도 고려해보세요.

eeeesong replied:
피드백에 따라 개선하였습니다 :)

napster-x-90 commented on ios/sidedish/sidedish/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    init(baseAddress: String){
+        self.requestManager = RequestManager(baseAddress: baseAddress)
+        self.urlSession = URLSession(configuration: .default)
```
`self.urlSession`을 사용하는데가 없어보이네요? 😭

eeeesong replied:
앗 그러네요🥲 수정을 거치다가 놓친 부분 같습니다. 수정하겠습니다!

napster-x-90 commented on ios/sidedish/sidedish/Network/RequestManager.swift:
```diff
@@ -0,0 +1,33 @@
+
+import Foundation
+
+class RequestManager {
```
상속을 하지 않을거라면 class 타입은 final을 명시해주세요.
:) 조금의 컴파일 성능 항샹이 있습니다.

eeeesong replied:
컴파일 성능 향상이 있는 줄은 몰랐네요!
앞으로 final을 꼭 명시해주도록 하겠습니다. 감사합니다 :)

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,54 @@
+        super.init(coder: coder)
+    }
+
+    func bind() {
```
private이 필요해보입니다.

napster-x-90 commented on ios/sidedish/sidedish/Presentation/MainMenuList/ViewModel/MenuCellViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+        turnonAppUsecase.manufactureForMainView()
+            .sink(receiveCompletion: { (result) in
+                switch result {
+                case .finished: print("finished")
```
print는 PR를 보내는 코드에 있으면 안됩니다. 불필요한 코드는 PR전에 제거해주세요 😃
에러는 어떻게 처리하면 좋을까요? 단순히 print로만 하면 안되겠죠?
유저에게 어떤 에러가 발생했고 지금 어떤 상태인지 알려줘야 할 의무가 있습니다.

eeeesong replied:
유저에게 alert가 갈 수 있도록 구현하였습니다.

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,54 @@
+        super.loadView()
+        self.menuTableView.dataSource = tableViewDataSource
+        self.menuTableView.delegate = tableViewDelegate
+        mainMenuViewModel.configureMainmenuBoard()
```
물론 실제 환경에서는 그럴일이 없겠지만,
bind전에 configureMain Menu Board의 응답이 먼저 오게되면 해당 스트림이 발생을 하나요?
Cold Observable이라면 응답이 오지 않을 것 같습니다.

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainTableViewDataSource.swift:
```diff
@@ -0,0 +1,32 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell : MenuCell = tableView.dequeueReusableCell(withIdentifier: "MenuCell", for: indexPath) as? MenuCell else { return UITableViewCell() }
```
MenuCell의 대한 Identifier는 MenuCell에 직접 static으로 사용하는게 좀더 좋아보입니다. 😀

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainTableViewDataSource.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell : MenuCell = tableView.dequeueReusableCell(withIdentifier: "MenuCell", for: indexPath) as? MenuCell else { return UITableViewCell() }
+        cell.updateMenu(titleText: menuCellViewModel.dishes.body[indexPath.section].sideDish(at: indexPath.row)!.title)
```
ForcedUnWrapping 은 굉장히 위험한 코드입니다. 다르게 처리해보세요 😄

eeeesong replied:
이 부분은 실제 API를 적용하는 과정에서 수정해보도록 하겠습니다 :)

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainTableViewDelegate.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        let headerView = UIView.init(frame: CGRect.init(x: 0, y: 0, width: tableView.frame.width, height: 32))
```
headerView만의 새로운 클래스로 만들어 따로 주입해 주는게 더 좋아보입니다.
비록 단순한 View일지라도 나중에 헤더가 복잡해지면 따로 분리해두면 수정하기 용이해집니다.

eeeesong replied:
새로운 클래스를 만들어서 생성하도록 하였습니다.

Settpark committed:
feat: Toast 기능 개발 : MenuCategory 부분

napster-x-90 reviewed (commented):
수고하셨습니다. :) 벌써부터 Clean Architecture를 적용하시다니 대단합니다.
다만 왜 Clean Architecture가 필요한지 꼭 이유를 알고 사용하셔야합니다.
그냥 남들이 쓰니까, 배웠으니까는 타당한 이유가 되지 않습니다.
😌

napster-x-90 reviewed (commented):
안녕하세요. 반갑습니다. Napster라고 합니다.
부족하지만 저도 열심히 한번 리뷰해볼께요!!

살펴보니 여러명이서 작성한 코드인게 눈에 잘 보이더라구요. 이렇게 잘 보인다는건 코드 컨벤션이 전혀 없다는 뜻이고, 다른사람이 읽을때 매우 힘든 코드라는 점입니다. 팀을 이루고 함께 프로젝트를 만들어가면 여러사람이 봐도 보기 좋아야겠죠?

Swift Lint를 해당 프로젝트에 적용해보세요. 최소한의 컨벤션은 지킬 수 있을꺼에요.
더 나아가서 현재 팀원들과 함께 Coding Style를 만들어서 지켜보는 건 어떨까요? 나중에 회사에 가면 회사마다 코딩 스타일이 있고 거기에 맞게 작성을 해야합니다. 만약 없다면 직접 만들어야겠죠? 


# 질문에 대한 답변
### Q.
1. UseCase를 다루는 형태를 고민했습니다. UseCase를 알고 있어야 하는 대상의 구분, UseCase의 역할 등을 고민하였고
현재 UseCase가 잘 쓰이고 있는 지에 대한 고민이 가장 컸습니다. -> 완전히 해결하지 못한 상태로 남아있습니다.

### A.
UseCase의 사용은 적절해보입니다.  다만 Entity와 UseCase가 강한 결합도가 있는게 아쉽습니다.
현재 `func manufactureForMainView() -> AnyPublisher<SideDishesCollection, Error>` 
는 SideDishesCollection를 서버에서 받은 그대로 사용하고 있는데,  SideDishes는 언제든 바뀔 수 있습니다. `SideDishFindable` 만들어서 결합도가 느슨해진줄 알았는데, `SideDishesCollection`의 body를 직접 ViewModel에서 사용하다보니 그렇게 되지 않게 되었습니다.
해당 앱의 Main에서 사용할때 느스한 결합도를 가진 적절한 Entities로 바꿔보세요. 
지금은 SideDishFindable로 의존성을 느슨하게 해주었지만 sideDishesCollection에서 구체화된 객체를 선언해서 쓰고있으니 해당
이점이 사라져 버렸습니다. 

### Q.
2. 현재 ViewModel에 sideDishesCollection의 구체 타입을 알고 있습니다. 구조상 뷰모델에서 도메인의 entities를 알고 있는 것에 대한 고민이 있는 데, 이 것을 단지 프로토콜로 전환한다고 해서 의존성이 끊어지는 지에 대한 의문이 남아 현재 진행하지 않고 구체 타입을 알고 있습니다. 프로토콜로 개선하면 이 문제를 해결할 수 있을까요??

### A.
1번에 대한 답이 어느정도 여기에도 포함되어 보이네요.
객체들간의 의존성을 완전히 끊을 수는 없습니다. 객체지향프로그래밍은 객체들간의 협력을 통해 이뤄지고 있으니까요. 의존은 필연적으로 발생하게 됩니다. 다만 어느 정도의 결합도를 가졌느나가 중요합니다.
객체들간의 의존을 프로토콜로 개선하면 결합도를 느슨하게 할 수 있습니다. 😙

### Q.
3. Network 모델 설계 시 재사용성을 고민하였습니다. Network 객체의 url을 외부에서 지정해줄 수 있도록 구현하고, Request 생성 객체 메소드에서 Enum 타입의 RequestType과 EndPoint를 넘겨줄 수 있도록 했습니다. EndPoint의 경우 hash 값 등 반찬 cell 선택에 따른 값 설정이 필요할 수 있어서 추후 API 형식에 따른 타입 변경이 필요해보입니다.

### A.
Moya를 한번 참고해보세요. :) 
https://moya.github.io

### Q.
4. Repository의 역할을 구분하는 기준이 명확하지 않아 현재의 프로젝트에서 repository의 영역을 구현하지 못했습니다. 혹시 현재 프로젝트에서 어떠한 방식으로 repository를 구현해야 하는지, 방향성에 대해서 질문 드리고 싶습니다.

### A.
현재 Use Case인 `TurnonAppUsecase`에 NetworkManager만 있는데 외부에서 Repository를 주입해줘야겠지요?
그렇게되면 NetworkManager의 구체를 사용하는게 아닌 Repository를 사용하게 될것입니다. 

Repository를 사용하는 이유는 무엇일까요? (한번 생각해보세요)

앱에서 사용하는 데이터가 API, Local Storage 등등 을 구분하지 않고 UserCase는 Repository에 메서드들만 사용하여 해당 데이터를 얻기만 하면 됩니다.  그러면 Repository가 NetworkManager를 의존하는 구조로 바뀌어야 할 것 같습니다.

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainTableViewDelegate.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return 56 // headerView의 height(32) + figma에서 지정해 준 헤더와 셀 간의 간격!
```
상수값으로 따로 선언해서 관리하는게 좋아보입니다.
직접적인 값의 사용은 좋지 않습니다 🥺

eeeesong committed:
feat: Toaster 기능 개발 Detail view 부분 (#10)

eeeesong commented:
안녕하세요 Napster! 저는 Lollo 입니다.
우선 꼼꼼하게 리뷰 해주셔서 정말 감사드립니다☺️
그리고 저희의 고민과 질문에 대해서도 너무 상세하게 잘 답변을 해주셔서 큰 도움이 될 것 같습니다!

저의 경우 이번 프로젝트에서는 Network에서 뿐만아니라 Core Data와 같은 Storage를 사용하게 되어서
Use Case, Repository 등 여러 계층을 구조적으로 잘 관리하기 위해 자연스럽게 Clean Architecture에 눈이 갔던 것 같습니다 :)

오늘 리뷰해주신 부분들 내일까지 ZG와 함께 한번 더 꼼꼼하게 확인해서 수정을 거친 뒤
다시 이곳에 push 해보도록 하겠습니다.

정말 감사합니다!

napster-x-90 commented:
화이팅 하시죠!!
저도 최대한 열심히 리뷰해볼께요! 부족하지만 감사합니다. 😄

eeeesong committed:
feat: Toast UI 개선

- CustomToaster: 각 토스트의 생성 메소드 분리하여 UI 정교화
- Pods-ToastView: Bubble의 Alignment를 설정할 수 있도록 변경

Settpark committed:
fix : 피드백 반영 (#30)
dquote>
dquote> class -> struct로 변경
dquote> 변수 var에서 let으로 변경
dquote> bind private 설정, Cold Observalbe 문제 해결
dquote> print문 제거 사용자에게 alertMessage 뜨도록 수정

Settpark committed:
fix : 피드백 반영 (#30)
dquote>
dquote> colleciton에 해당하는 protocol선언 (임시)
dquote> TablecustomHeaderView 선언

Settpark committed:
Merge branch 'feature/Toast' of https://github.com/ghojeong/sidedish into fix/firstPR

eeeesong committed:
fix: PR 피드백 반영 (#30)

- NetworkManager: Mock 객체 초기화 시에도 URL 외부 주입, 쓰지 않는 URLSession 삭제
- SideDish: CodingKey 추가

eeeesong committed:
Merge pull request #35 from ghojeong/fix/firstPR

[iOS] 1차 PR 수정

eeeesong commented:
안녕하세요 Napster, 좋은 아침입니다☀️
어제 주신 피드백 바탕으로 수정을 마쳤습니다 :) 
다음 목록을 수정했고, 코멘트 달아주신 부분에도 추가 코멘트 달아두었습니다! 

- [x]  SideDish: Class -> Struct, var -> let
- [x]  NetworkManager: Mock 객체 초기화 시에도 url 외부 주입, URLSession 부분 수정
- [x]  MainViewController: bind를 private하게, Cold Observable 문제
- [x]  MenuCellViewModel: print문 제거 -> 사용자에게 문제 상황 알리기
- [x]  MainTableViewDataSource: cell 이름 static 상수로 변경
- [ ]  MainTableViewDataSource: 강제 언래핑 제거 ➡️ `실제 API를 적용하는 과정에서 수정해보겠습니다`
- [x]  MainTableViewDelegate: headerView 분리하여 생성, 높이 값 상수로 저장

문제가 없으시다면 머지 부탁드리겠습니다 :) 감사합니다!

napster-x-90 commented on ios/sidedish/sidedish/Presentation/MainMenuList/ViewModel/MenuCellViewModel.swift:
```diff
@@ -52,6 +52,13 @@ class MenuCellViewModel{
-        
+    }
+    
+    func configureAlertMessage(completion : @escaping ((String)->())){
```
이왕 Combine을 쓰고있으면 completion handler말고 publisher로 해결해보는건 어떨까요?

eeeesong replied:
error alert 처리도 publisher로 바꾸었습니다 :)

napster-x-90 commented on ios/sidedish/sidedish/Presentation/ViewController/MainViewController.swift:
```diff
@@ -40,20 +40,30 @@ class MainViewController: UIViewController, ViewChangable {
+    }
+    
+    private func showAlert(message : String){
+        DispatchQueue.main.async {
```
캡쳐리스트는 꼭 필요합니다.
강한참조가 발생할 수 있거든요 😆

napster-x-90 reviewed (commented):
몇가지만 조금 수정하면 될 것 같습니다.

Settpark committed:
fix: API 수정하고 내부 구조 변경

ViewModel에서 카테고리, 제목 따로따로 받아서 화면에 출력함

Settpark committed:
fix: 이미지까지 TableView에 로드 (#36)

Datasource의 38번 줄 Data 사용한 부분 수정 요망

eeeesong committed:
fix: 반찬Section GET 메소드 추가(#36)

eeeesong committed:
refactor: PR 2차 피드백 반영 & 반찬 목록 GET 개선

eeeesong committed:
chore: 주석 삭제, 메소드 순서 변경

Settpark committed:
Merge pull request #40 from ghojeong/fix/ChangeAPI

고생 많았어요 :)

eeeesong commented:
안녕하세요 Napster :) 수정이 조금 늦어졌습니다. 더불어 일요일에 코멘트를 남기게 되어 죄송하다는 말씀 드립니다🥲

```swift
mainMenuViewModel.$errorMessage
             .receive(on: DispatchQueue.main)
             .sink(receiveCompletion: { _ in}, receiveValue:{ [weak self] error in self?.showAlert(message: error) })
             .store(in: &self.subscription)
```

위와 같이 수정을 했는데, 말씀하신 강한 참조 부분을 제대로 해결했는지 모르겠습니다ㅠㅠ
[weak self]로 작성한 부분으로 캡처 부분을 해결했다고 생각했는데, 괜찮으시다면 맞게 한 것인지 한번 확인 부탁드립니다!

napster-x-90 commented:
네 캡쳐 리스트를 통해 강한 참조 발생을 방지할 수 있습니다.
꼭! 어떨때, 어떻게 써야하는지 숙지하고 계셔야합니다.

capture list

eeeesong commented:
> 꼭! 어떨때, 어떻게 써야하는지 숙지하고 계셔야합니다.

넵 명심하고 익히겠습니다!
감사합니다 :)
